### DISCUSSION TITLE
Build on Apple Silicon + Homebrew

### DISCUSSION THREAD

**Discussion by sigmafelix** [Spatiotemporal-Exposures-and-Toxicology/PrestoGP_Pesticides#18] [General]
Although this is not recommended by the developer, we need to modify `./src/Makevars` to build `PrestoGP` on Apple Silicon with Homebrew. The settings below enables us to bind OpenMP and library paths that were set by Homebrew.

1. In `./src/Makevars`:
```
## With R 3.1.0 or later, you can uncomment the following line to tell R to 
## enable compilation with C++11 (where available)
##
## Also, OpenMP support in Armadillo prefers C++11 support. However, for wider
## availability of the package we do not yet enforce this here.  It is however
## recommended for client packages to set it.
##
## And with R 3.4.0, and RcppArmadillo 0.7.960.*, we turn C++11 on as OpenMP
## support within Armadillo prefers / requires it
CXX_STD = CXX11

PKG_CXXFLAGS = $(SHLIB_OPENMP_CXXFLAGS) 
PKG_LIBS = -L/opt/homebrew/Cellar/gcc/13.2.0/lib/gcc/13 -L/opt/homebrew/opt/libomp/lib $(SHLIB_OPENMP_CXXFLAGS) $(LAPACK_LIBS) $(BLAS_LIBS) 
PKG_CPPFLAGS = -DARMA_64BIT_WORD=1 -I/opt/homebrew/opt/libomp/include

```

2. Make a new file `${HOME}/.R/Makevars`
- We need to have this because the Makevars is applied after the immediate path binding `./src/Makevars`.
```
CPPFLAGS += -I/opt/homebrew/include -Xclang -fopenmp
LDFLAGS += -L/opt/homebrew/lib -lomp
```

Reference
- https://stackoverflow.com/questions/49217539/package-build-ignores-makevars-flags
- https://cran.r-project.org/doc/manuals/r-devel/R-exts.html#Interfacing-C_002b_002b-code
- https://github.com/Rdatatable/data.table/issues/5419

**Comment by kyle-messier** [comment]
@sigmafelix  Interesting, and you think this would open it up to GPU support? I thought there would need to be more edits in the source code (C++ and/or R) to do that.

**Comment by sigmafelix** [comment]
@Spatiotemporal-Exposures-and-Toxicology GPU support is a different avenue, I guess. As far as I understood, the entire C++ codebase needs to be converted to use CUDA API or we can develop a separate CUDA version alongside with the present C++ codebase to allow user flexibility (e.g., [R package GWModel](https://cran.r-project.org/web/packages/GWmodel/index.html)).

**Comment by sigmafelix** [comment]
It seems like PrestoGP uses RcppArmadillo, not OpenMP. Thus, the OpenMP part I mentioned above may be irrelevant.
